File: 2020/src/day1.jl
module Day1

include("InputHelper.jl")

function solve()
    input = get_input()

    println("Part one $(solve_part_one(input))")

    println("Part two $(solve_part_two(input))")
end

function get_input()::Array{Int,1}
    InputHelper.parse("1", line -> parse(Int, line))
end

function solve_part_one(input::Array{Int,1})
    for (i, j) in Iterators.product(input, input)
        if sums_to_2020(i, j)
            return i * j
        end
    end

    nothing
end


function solve_part_two(input::Array{Int,1})
    for (i, j, k) in Iterators.product(input, input, input)
        if sums_to_2020(i, j, k)
            return i * j * k
        end
    end

    nothing
end

sums_to_2020(xs::Int...) = sum(xs) == 2020
end

File: aoc-2022/src/day1.rs
use crate::parser::parser;


pub fn solve() {
    let input = parser::parse(1, "\n\n", parse_group);
    let one = solve1(input.to_vec());
    println!("Part one: {}", one);
    let two = solve2(input.to_vec());
    println!("Part two: {}", two);
}


fn solve1(input: Vec<Vec<i32>>) -> i32 {
    input.iter().map(|elf| elf.iter().sum()).max().unwrap()
}

fn solve2(input: Vec<Vec<i32>>) -> i32 {
    let mut elf_snacks: Vec<i32> = input.iter().map(|elf| elf.iter().sum()).collect();
    elf_snacks.sort();
    elf_snacks.iter().rev().take(3).sum()
}

fn parse_group(group: &str) -> Vec<i32> {
    group
        .lines()
        .filter(|l| l.len() > 0)
        .map(|l| l.parse::<i32>().unwrap())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part_one_test() {
        let input = parser::parse(1, "\n\n", parse_group);
        assert_eq!(solve1(input), 71924);
    }

    #[test]
    fn part_two_test() {
        let input = parser::parse(1, "\n\n", parse_group);
        assert_eq!(solve2(input), 210406);
    }
}
